### DISCUSSION TITLE
Can I use both first_name last_name  for  page title and global search ?

### DISCUSSION THREAD

**Discussion by sergeynilov** [filamentphp/filament#13303] [Help] [answered]
### Package

Other

### Package Version

^3.2-stable

### How can we help you?

In laravel 10 / filamentphp 3 app in Employee model I havea field defined :
```
    function getFullNameAttribute()
    {
        return ($this->attributes['first_name'] ?? '') . ' ' . ($this->attributes['last_name'] ?? '');
    }

```
But when in resource I define :
```
class EmployeeResource extends Resource
{
    protected static ?string $model = Employee::class;
    protected static ?string $recordTitleAttribute = 'first_name';
    ...

    public static function getGlobalSearchResultTitle(Model $record): string
    {
        return $record->full_name;
    }

```
I got error :

`    SQLSTATE[42S22]: Column not found: 1054 Unknown column 'full_name' in 'where clause'
`
as have no full_name field in employees table...

If there is a way to define both first_name last_name for page title and global search, as string $recordTitleAttribute is used in both cases...

**Comment by xalabama** [comment]
What about calling your full name as a public method from your model?  It feels like a bad practice to add or call attributes inside the model they doesn't exists in database table.

**Reply by sergeynilov** [reply]
Could you, please, give an example/more details how to implement this 

>     What about calling your full name as a public method from your model ? 

? I am not sure what you are about.

**Reply by xalabama** [reply]
I mean that you simply can do something like this:

In your model
```php
public function getFullName(): string
{
    $attributes = $this?->attributes ?? [];

    return data_get($attributes,  'first_name', '') . ' ' . data_get($attributes, 'last_name',''); 
}
```

And than call it from where you need it: 
```php
class EmployeeResource extends Resource
{
    protected static ?string $model = Employee::class;
    protected static ?string $recordTitleAttribute = 'first_name';
    ...

    public static function getGlobalSearchResultTitle(Model $record): string
    {
        return $record->getFullName();
    }
```

This should working.
But maybe i have misunderstood you.

**Reply by christmex** [reply]
@xalabama 👆👆👆👆 workssss!!

**Reply by sergeynilov** [reply]
That it really works !

But I also woulld like to see douibled name in page title and as I can not call model method in  desclaration :
`    protected static ?string $recordTitleAttribute = 'first_name';
`
I try to overwrite method :

 ```
   public static function getRecordTitleAttribute(): ?string
    {
        $record->getFullName();
    }

```
But as this method has no $record var I tried to make it like :+1: 

```
    public static function getRecordTitleAttribute(): ?string
    {
        return self::getRecord()->getFullName();
    }

```
Got error  : 
`Method App\Filament\App\Resources\EmployeeResource::getRecord does not exist.
`
How do it ?

**Reply by xalabama** [reply]
Yes for sure, because you are try to call a dynamic method inside a static one. 

Are you already familiar with the concept of Mutators and Accessors of laravels eloquent?  If not you can have a look here first:
https://laravel.com/docs/11.x/eloquent-mutators#defining-an-accessor

This is also referenced inside the filament documentation, here: 
https://filamentphp.com/docs/3.x/panels/resources/getting-started#record-titles

So this means you should be able to define a custom accessor inside your model. Previous we have created a method called 'getFullName', that method can reused here. 

```php
protected function recordTitle(): \Illuminate\Database\Eloquent\Casts\Attribute
{
    return Attribute::make(
        get: fn (mixed $value) => $this->getFullName()
    );
}
```

Inside your resource class, you can now set the accessor: 

```php
protected static ?string $recordTitleAttribute = 'recordTitle';
```

Did this work for you?

**Reply by sergeynilov** [reply]
Thanks!
@xalabama I remade as you wrote and in page title I see valid title of 2 fields

But when I tried to run global search I got error :
```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'recordTitle' in 'where clause'
    SELECT
    *
    FROM
    `employees`
    WHERE
    `employees`.`branch_id` IN (1)
    AND (`recordTitle` LIKE % judy %)
    limit

```
?

**Reply by xalabama** [reply, marked as answer]
Okay my fault, sorry. 

Lets do this step by step. 

I think you need to refactor your table structure for this, because you want to make your first_name and last_name attributes searchable. If you bagging them inside a json column its hard to search for them, because than you need to modify the whole global search query and work with OPENJSON. 
And a search without sql will result an app that consumes to much resources for searching. 

So at first create two new columns inside your employees table for first_name and last_name. Don't forget to modify your model also, specially the getFullName method. 

And modify some methods inside the resource: 

```php
// This will modify the title of items in search results list
public static function getGlobalSearchResultTitle(Model $record): string
{
    return $record->getFullName();
}

// This will allows you to define attributes they can searched for in the global search
public static function getGloballySearchableAttributes(): array
{
    return [
        'first_name', 
        'last_name', 
        // ...
    ];
}
```

Maybe this way works for you?

**Reply by sergeynilov** [reply]
Sorry, if I badly put it, but Employee model class from start has both "first_name" and "last_name" fields and I do not want to change.
So I hardly understand message above.

**Reply by xalabama** [reply]
I am sorry that you are not happy with the solutions offered. I assumed that your fields are stored in a JSON column, hence my suggestion to move them to indexable columns. If you find a better solution I would be happy if you could share it here. Otherwise, I don't know what to do for now and can't give you any more input.

However, I wish you every success in finding a suitable solution.

**Comment by christmex** [comment]
Hi @sergeynilov there is a way, to use concat in SQL ofc, but we need to set Virtual Column in the database. I'll try it first, and back to you if it's working.


Edit:
After trying to use this approach, there is something that I don't like
1. if you don't set one of first_name or last_name, then it won't work, the virtual column works if you set both in other words the user should have first_name and last_name if no, this won't work.
2. only support MySQL and SQLite.

I think @xalabama ways better, thanks @xalabama sometimes I need something like this also.
